test/extensions/filters/network/ssh/service_userauth_test.cc

```diff
@@ -0,0 +1,746 @@
+                                 data).status());
+  wire::UserAuthRequestMsg user_auth;
+  ASSERT_OK(user_auth.decode(buf, buf.length()).status());
+
+  auto r = service_->handleMessage(wire::Message{user_auth});
+  ASSERT_EQ(r.code(), absl::StatusCode::kUnimplemented);
+  ASSERT_EQ(r.message(), "unknown or unsupported auth method");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowUpstream) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto allow = msg->mutable_auth_response()->mutable_allow();
+  auto upstream = allow->mutable_upstream();
+  upstream->set_hostname("example-hostname");
+  EXPECT_CALL(*transport_, streamId())
+    .WillOnce(Return(42));
+  wire::ExtInfoMsg ext_info;
+  wire::test::populateFields(ext_info);
+  EXPECT_CALL(*transport_, peerExtInfo())
+    .WillOnce(Return(ext_info));
+
+  AuthStateSharedPtr state;
+  EXPECT_CALL(*transport_, initUpstream(_))
+    .WillOnce(SaveArg<0>(&state));
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+
+  ASSERT_THAT(*state->allow_response, Envoy::ProtoEq(*allow));
+  ASSERT_EQ(42, state->stream_id);
+  ASSERT_EQ(*state->downstream_ext_info, ext_info);
+  ASSERT_EQ(ChannelMode::Normal, state->channel_mode);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowInternal) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto allow = msg->mutable_auth_response()->mutable_allow();
+  auto internal = allow->mutable_internal();
+  ProtobufWkt::Value v;
+  v.set_string_value("example-metadata-value");
+  ProtobufWkt::Struct metadata_struct{};
+  (*metadata_struct.mutable_fields())["example-metadata-key"] = v;
+  auto filter_metadata = *internal->mutable_set_metadata()->mutable_filter_metadata();
+  filter_metadata["example-filter-name"] = metadata_struct;
+
+  EXPECT_CALL(*transport_, streamId())
+    .WillOnce(Return(42));
+  wire::ExtInfoMsg ext_info;
+  wire::test::populateFields(ext_info);
+  EXPECT_CALL(*transport_, peerExtInfo())
+    .WillOnce(Return(ext_info));
+
+  AuthStateSharedPtr state;
+  EXPECT_CALL(*transport_, initUpstream(_))
+    .WillOnce(SaveArg<0>(&state));
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+
+  ASSERT_THAT(*state->allow_response, Envoy::ProtoEq(*allow));
+  ASSERT_EQ(42, state->stream_id);
+  ASSERT_EQ(*state->downstream_ext_info, ext_info);
+  ASSERT_EQ(ChannelMode::Hijacked, state->channel_mode);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowUnsupportedTarget) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  msg->mutable_auth_response()->mutable_allow();
+  EXPECT_CALL(*transport_, streamId())
+    .WillOnce(Return(42));
+  EXPECT_CALL(*transport_, peerExtInfo())
+    .WillOnce(Return(wire::ExtInfoMsg{}));
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_EQ(r.code(), absl::StatusCode::kInternal);
+  ASSERT_EQ(r.message(), "invalid target");
```

Comment:
```suggestion
  ASSERT_EQ(absl::InternalError("invalid target"), service_->handleMessage(std::move(msg)));
```
There are other instances in this file where this simplification could be applied, but I won't add separate comments for each.